File: crates/gui/frontend/src/hooks/licenses.ts
/**
 * What the app can say about its own licensing.
 *
 * Three documents live behind these calls: Hydra's own licence, the
 * commercial-licence offer, and the notices of every open-source package
 * this build ships. All three are embedded in the binary — see
 * `crates/gui/src/commands/about.rs` — so they are readable on a machine
 * that has never seen the repository.
 *
 * The notices are split deliberately. The component list is small enough to
 * fetch whole; the licence texts are a megabyte between them and are
 * fetched one at a time, by the index a component names.
 */

import { tryInvoke, tryInvokeOr } from "./ipc";

/** Hydra's own licensing. */
export interface LicenseInfo {
  /** SPDX identifier — `AGPL-3.0-only`. */
  spdx: string;
  /** The full licence text. */
  text: string;
  /** The commercial-licence document, in Markdown. */
  commercial: string;
  /** Where the corresponding source is published. */
  sourceUrl: string;
}

/** One licence file a package ships. */
export interface NoticeFile {
  /** The file's name in the package (`LICENSE`, `LICENSE-APACHE`, …). */
  name: string;
  /** Index into the shared text pool, for `getThirdPartyLicenseText`. */
  text: number;
}

/** One open-source package this build ships. */
export interface ThirdPartyComponent {
  name: string;
  version: string;
  /** `"rust"` or `"npm"`. */
  ecosystem: string;
  /** The declared SPDX expression; empty when the package declares none. */
  spdx: string;
  /** The package's own home; empty when it states none. */
  url: string;
  files: NoticeFile[];
}

export async function getLicenseInfo(): Promise<LicenseInfo | null> {
  return tryInvoke<LicenseInfo>("get_license_info");
}

export async function listThirdPartyComponents(): Promise<
  ThirdPartyComponent[]
> {
  return tryInvokeOr<ThirdPartyComponent[]>(
    "list_third_party_components",
    undefined,
    [],
  );
}

export async function getThirdPartyLicenseText(
  index: number,
): Promise<string | null> {
  return tryInvoke<string>("get_third_party_license_text", { index });
}

// ── Decisions ─────────────────────────────────────────────────────────────────

/**
 * Whether a component answers to what was typed.
 *
 * Matched against the version too, which is not obvious but is the reason
 * the box is there: the question a reader brings to a notices list is
 * almost always "is *this* package in here, and which version", arriving
 * from an advisory that names both.
 */
export function matchesComponentQuery(
  component: ThirdPartyComponent,
  query: string,
): boolean {
  const q = query.trim().toLowerCase();
  if (!q) return true;
  return [
    component.name,
    component.version,
    component.spdx,
    component.ecosystem,
  ].some((field) => field.toLowerCase().includes(q));
}

/**
 * What stands in for a licence text when the package ships none.
 *
 * Around eighty do: they declare a licence in their manifest and leave it
 * at that. Their row is the whole notice they get, so it has to say what
 * the licence is and where the package lives rather than showing an empty
 * panel that reads as a missing notice.
 *
 * Returns null when the package ships its own text, which is the normal
 * case and needs no substitute.
 */
export function noticeFallback(component: ThirdPartyComponent): string | null {
  if (component.files.length > 0) return null;
  const licence = component.spdx
    ? `Declared as ${component.spdx}.`
    : "No licence declared.";
  return component.url
    ? `${licence} This package ships no licence file; see ${component.url}.`
    : `${licence} This package ships no licence file.`;
}

/**
 * A link in an embedded repository document, resolved against the source.
 *
 * The commercial-licence document is the repository's own Markdown, and it
 * points at sibling files — `[the AGPL](LICENSE)`. Read from the
 * repository that resolves; read from inside an installed app there is no
 * sibling to resolve to, and the link opens nothing at all. Rewriting
 * relative targets onto the published source is what makes the document
 * still work once it has left the tree it was written in.
 */
export function repoLink(href: string, sourceUrl: string): string {
  if (!href) return href;
  // Already somewhere, or going nowhere: absolute URLs, mail links and
  // in-document anchors are all left exactly as written.
  if (/^[a-z][a-z0-9+.-]*:/i.test(href) || href.startsWith("#")) return href;
  if (href.startsWith("//")) return href;
  const path = href.replace(/^\.\//, "").replace(/^\//, "");
  return `${sourceUrl.replace(/\/+$/, "")}/blob/main/${path}`;
}

File: scripts/licenses.py
#!/usr/bin/env python3
"""Collect the licence notices of everything the desktop app ships.

Hydra's own licence is one file in this repository. The four hundred-odd
open-source packages the app is built from are not: MIT, BSD and Apache-2.0
all ask that their copyright notice travel with the binary, and a notice
that only exists in a `~/.cargo` cache on a build machine has not travelled
anywhere.

So this script walks what the GUI actually links — the normal (not dev, not
build) dependency graph of `hydra-gui`, plus the frontend's production npm
tree — reads the licence text out of each package, and writes one JSON
document the app embeds and shows under Settings → About.

Two decisions are worth stating, because both are visible in the output:

*Every platform's dependencies, not this one's.* `cargo metadata` is read
unfiltered, so the Windows and Linux halves of the tree are included on a
macOS machine. We ship all three, and a notices file that changes depending
on who regenerated it would be worse than useless as a `--check` gate.

*Texts are shared, components are not.* Hundreds of packages carry the same
Apache-2.0 text; MIT texts differ by a single copyright line and so mostly
do not. Identical texts are stored once and referenced by index, which is a
size decision only — every component still names its own licence, and a
package shipping two files keeps them as two, so the boilerplate half of a
dual licence dedupes even where the copyright half cannot.

Usage:
    scripts/licenses.py            regenerate the notices file
    scripts/licenses.py --check    fail if the committed file is stale
"""

import hashlib
import json
import pathlib
import re
import subprocess
import sys

ROOT = pathlib.Path(__file__).resolve().parents[1]
OUT = ROOT / "crates" / "gui" / "resources" / "third-party-licenses.json"
FRONTEND = ROOT / "crates" / "gui" / "frontend"

# The crate whose dependency graph is "what the app ships". The GUI links
# the engines, the report layer and the SDK, so starting here reaches
# everything without listing it.
ROOT_CRATE = "hydra-gui"

# Files a package might hold its licence in. Matched case-insensitively
# against the whole name: LICENSE, LICENSE.md, LICENSE-APACHE, licence.txt,
# COPYING, NOTICE, UNLICENSE.
#
# The extension list is closed, and the qualifier before it may not contain
# a dot. Allowing either to be `.anything` let `license_key.rs` through —
# a source file whose name begins with the word, whose contents are not a
# notice, and which would have been embedded in the app as one.
LICENSE_FILE_RE = re.compile(
    r"^(LICEN[CS]E|COPYING|COPYRIGHT|NOTICE|UNLICEN[CS]E)"
    r"([-_][A-Za-z0-9-]+)*(\.(md|txt|rst))?$",
    re.IGNORECASE,
)

# A licence text is a page or two. Anything past this is a vendored corpus
# that happens to match the name pattern, and embedding it would bloat the
# binary for nothing.
MAX_TEXT_BYTES = 200_000


# ── Selection and normalisation (pure) ────────────────────────────────────────


def pick_license_files(names):
    """The licence-bearing file names among `names`, in a stable order.

    Sorted rather than left in directory order: the same package must
    produce the same text on every machine, and `os.listdir` promises
    nothing about order.
    """
    return sorted(n for n in names if LICENSE_FILE_RE.match(n))


def normalise_text(text: str) -> str:
    """The form two texts are compared in.

    Line endings and trailing whitespace differ between packages carrying
    what is otherwise the same licence — CRLF from a Windows contributor,
    a stray space before a newline. Neither changes what the licence says,
    and treating them as different texts would store the Apache-2.0 licence
    a hundred times over.
    """
    lines = text.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    return "\n".join(line.rstrip() for line in lines).strip() + "\n"


def dedupe(components):
    """Split components into a shared text pool and index references.

    Returns `(texts, components)` where each component's `files` is a list
    of `[file name, index into texts]` pairs — empty when the package
    shipped no licence file at all.

    Files stay separate rather than being concatenated per package. A
    dual-licensed crate ships LICENSE-APACHE beside LICENSE-MIT, and the
    Apache half is the same eleven kilobytes in every one of them while the
    MIT half carries a copyright line that is not. Joined, each pair is a
    unique text and the boilerplate is stored hundreds of times; kept
    apart, it is stored once.
    """
    texts = []
    index_of = {}
    out = []
    for c in components:
        files = []
        for name, body in c.get("files", []):
            key = hashlib.sha256(body.encode("utf-8")).hexdigest()
            if key not in index_of:
                index_of[key] = len(texts)
                texts.append(body)
            files.append([name, index_of[key]])
        out.append({**c, "files": files})
    return texts, out


def sort_key(component):
    """Alphabetical by name, then by version, ecosystem last.

    Version is compared as a string on purpose: this orders the file, it
    does not decide anything, and a semver parse here would be a second
    place for semver to be wrong.
    """
    return (component["name"].lower(), component["version"], component["ecosystem"])


# ── Collection (pure, given the tool output) ──────────────────────────────────


def rust_components(metadata):
    """Every crate the GUI links, from `cargo metadata` output.

    Walks the resolve graph from `hydra-gui` following *normal* dependency
    edges only. Dev-dependencies are test-time and build-dependencies run
    on the build machine; neither is inside the binary we hand to anyone,
    and including them would put notices in the file for code nobody
    receives.
    """
    packages = {p["id"]: p for p in metadata["packages"]}
    nodes = {n["id"]: n for n in metadata["resolve"]["nodes"]}
    roots = [p["id"] for p in metadata["packages"] if p["name"] == ROOT_CRATE]
    if not roots:
        raise SystemExit(f"error: {ROOT_CRATE} not found in cargo metadata")

    seen = set()
    stack = list(roots)
    while stack:
        pid = stack.pop()
        if pid in seen:
            continue
        seen.add(pid)
        for dep in nodes[pid]["deps"]:
            # `kind: null` is a normal dependency; "dev" and "build" are not.
            if any(k.get("kind") is None for k in dep["dep_kinds"]):
                stack.append(dep["pkg"])

    out = []
    for pid in seen:
        pkg = packages[pid]
        # A workspace member is Hydra's own code, licensed by the file the
        # About panel shows separately.
        if pkg.get("source") is None:
            continue
        out.append(
            {
                "name": pkg["name"],
                "version": pkg["version"],
                "ecosystem": "rust",
                "spdx": pkg.get("license") or "",
                "url": pkg.get("repository") or "",
                "dir": str(pathlib.Path(pkg["manifest_path"]).parent),
            }
        )
    return out


def npm_components(listing):
    """Every production npm package, from `pnpm licenses list --prod --json`.

    pnpm groups by licence and lists a package once with parallel
    `versions` and `paths` arrays. They are zipped rather than assumed
    equal in length: a mismatch there would silently attribute one
    version's notice to another.
    """
    out = []
    for entries in listing.values():
        for entry in entries:
            versions = entry.get("versions") or []
            paths = entry.get("paths") or []
            for version, path in zip(versions, paths):
                out.append(
                    {
                        "name": entry["name"],
                        "version": version,
                        "ecosystem": "npm",
                        "spdx": entry.get("license") or "",
                        "url": entry.get("homepage") or "",
                        "dir": path,
                    }
                )
    return out


# ── Reading the texts (filesystem) ────────────────────────────────────────────


def read_license_files(directory: str):
    """The licence files a package ships, as `(name, normalised text)`."""
    d = pathlib.Path(directory)
    if not d.is_dir():
        return []
    named = []
    for name in pick_license_files(p.name for p in d.iterdir() if p.is_file()):
        f = d / name
        if f.stat().st_size > MAX_TEXT_BYTES:
            continue
        try:
            named.append((name, normalise_text(f.read_text(encoding="utf-8", errors="replace"))))
        except OSError:
            continue
    return named


def run_json(cmd, cwd=None):
    proc = subprocess.run(cmd, cwd=cwd, capture_output=True, text=True)
    if proc.returncode != 0:
        raise SystemExit(f"error: {' '.join(cmd)} failed:\n{proc.stderr.strip()}")
    return json.loads(proc.stdout)


def collect():
    metadata = run_json(["cargo", "metadata", "--format-version", "1"], cwd=ROOT)
    listing = run_json(["pnpm", "licenses", "list", "--prod", "--json"], cwd=FRONTEND)

    components = rust_components(metadata) + npm_components(listing)

    # A package whose directory is not there reads as one shipping no
    # licence file, which is a wrong notice rather than a missing one — and
    # under `--check` it would fail as unexplained drift. Say what happened
    # instead.
    absent = [c for c in components if not pathlib.Path(c["dir"]).is_dir()]
    if absent:
        names = ", ".join(f"{c['name']} {c['version']}" for c in absent[:5])
        raise SystemExit(
            f"error: {len(absent)} package directories are missing ({names}…). "
            f"Run `cargo fetch` and `pnpm install` before generating notices."
        )

    for c in components:
        c["files"] = read_license_files(c.pop("dir"))
    components.sort(key=sort_key)

    texts, components = dedupe(components)
    return {
        "counts": {
            "rust": sum(1 for c in components if c["ecosystem"] == "rust"),
            "npm": sum(1 for c in components if c["ecosystem"] == "npm"),
        },
        "texts": texts,
        "components": components,
    }


def serialise(document) -> str:
    # Compact separators: this file is generated, read by a program, and
    # embedded in the binary — the whitespace of a pretty print is a
    # hundred kilobytes of nothing.
    return json.dumps(document, ensure_ascii=False, separators=(",", ":"), sort_keys=True) + "\n"


def main():
    check = "--check" in sys.argv[1:]
    document = serialise(collect())

    if check:
        current = OUT.read_text(encoding="utf-8") if OUT.exists() else ""
        if current != document:
            raise SystemExit(
                f"error: {OUT.relative_to(ROOT)} is out of date — run `just licenses` "
                f"and commit the result."
            )
        print(f"{OUT.relative_to(ROOT)} is up to date.")
        return

    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(document, encoding="utf-8")
    counts = json.loads(document)["counts"]
    print(
        f"Wrote {OUT.relative_to(ROOT)} — {counts['rust']} crates, "
        f"{counts['npm']} npm packages, {len(json.loads(document)['texts'])} distinct texts."
    )


if __name__ == "__main__":
    main()
